karapace/schema_models.py

```diff
@@ -116,20 +140,71 @@ def normalize_schema_str(schema_str: str, schema_type: SchemaType) -> str:
         return schema_str
 
     def __str__(self) -> str:
-        return self.schema_str
+        if self.schema_type == SchemaType.PROTOBUF:
+            return self.schema_str
+
+        if self._str_cached is None:
+            self._str_cached = json_encode(self.to_dict())
+        return self._str_cached
 
     def __repr__(self) -> str:
         return f"TypedSchema(type={self.schema_type}, schema={str(self)})"
 
+    @property
+    def schema(self) -> Union[Draft7Validator, AvroSchema, ProtobufSchema]:
+        if self._schema_cached is not None:
+            return self._schema_cached
+        if self.schema_type is SchemaType.AVRO:
+            try:
+                self._schema_cached = parse_avro_schema_definition(self.schema_str)
+            except (SchemaParseException, json.JSONDecodeError, TypeError) as e:
+                raise InvalidSchema from e
+
+        elif self.schema_type is SchemaType.JSONSCHEMA:
+            try:
+                self._schema_cached = parse_jsonschema_definition(self.schema_str)
+                # TypeError - Raised when the user forgets to encode the schema as a string.
+            except (TypeError, json.JSONDecodeError, SchemaError, AssertionError) as e:
+                raise InvalidSchema from e
+
+        elif self.schema_type is SchemaType.PROTOBUF:
+            try:
+                self._schema_cached = parse_protobuf_schema_definition(self.schema_str, self.references, self.dependencies)
+            except (
+                TypeError,
+                SchemaError,
+                AssertionError,
+                ProtobufParserRuntimeException,
+                IllegalStateException,
+                IllegalArgumentException,
+                ProtobufError,
+                ProtobufException,
+                ProtobufSchemaParseException,
+            ) as e:
+                raise InvalidSchema from e
+        else:
+            raise InvalidSchema(f"Unknown parser {self.schema_type} for {self.schema_str}")
+        return self._schema_cached
+
+    def get_references(self) -> Optional[List[Reference]]:
+        return self.references
+
     def __eq__(self, other: Any) -> bool:
-        return isinstance(other, TypedSchema) and str(self) == str(other) and self.schema_type is other.schema_type
+        return (
+            isinstance(other, (TypedSchema, ValidatedTypedSchema))
```

Comment:
```suggestion
            isinstance(other, TypedSchema)
```

TypedSchema is the base class for both.

Reply:
yes